File: AddTwoNumbers.py
"""
You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.

You may assume the two numbers do not contain any leading zero, except the number 0 itself.
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

def addTwoNumbers(self, l1, l2):
    """
    :type l1: ListNode
    :type l2: ListNode
    :rtype: ListNode
    """
    result = []
    head1 = l1
    head2 = l2
    carry = 0 
    while head1 and head2:
        #print("calculating sum = ",head1.val + head2.val)
        if head1.val + head2.val+carry <=9:
            result.append(head1.val + head2.val+carry)
            #print("result ->",result)
            carry = 0
        else:
            result.append((head1.val + head2.val +carry)%10)
            carry = (head1.val + head2.val + carry)//10
        head1 = head1.next
        head2 = head2.next
    while head1:
        if head1.val + carry <= 9:
            result.append(head1.val+carry)
            carry = 0
        else:
            result.append((head1.val+carry)%10)
            carry = (head1.val + carry)//10
        head1 = head1.next
    while head2:
        if head2.val + carry <= 9:
            result.append(head2.val+carry)
            carry = 0
        else:
            result.append((head2.val+carry)%10)
            carry = (head2.val + carry)//10
        head2 = head2.next
    if carry >0:
        result.append(carry)
    return result
    

File: AddTwoNumbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0;
        int num = 0;
        ListNode *answer = new ListNode(-1);
        ListNode *res= answer;
        while(l1!=NULL && l2!=NULL){
            num = l1->val + l2->val + carry;
            carry = num/10;
            num = num%10;
            res->next = new ListNode(num);
            l1 = l1->next;
            l2 = l2->next;
            res = res->next;
        }
        while(l1!=NULL){
            num = l1->val + carry;
            carry = num/10;
            num = num %10;
            res->next = new ListNode(num);
            l1 = l1->next;
            res = res->next;
        }
        while(l2!=NULL){
            num = l2->val + carry;
            carry = num/10;
            num = num %10;
            res->next = new ListNode(num);
            l2 = l2->next;
            res = res->next;
        }
        if(carry != 0){
            res->next = new ListNode(carry);
        }
        return answer->next;
    }
};
